### DISCUSSION TITLE
Tailwind: extend theme with a custom _responsive_ spacing

### DISCUSSION THREAD

**Discussion by ivan-kleshnin** [tailwindlabs/tailwindcss#9760] [Help] [answered]
Copied from my unaswered SO question https://stackoverflow.com/questions/74313306/tailwind-extend-theme-with-a-custom-responsive-spacing (in case someone's interested in extra karma there)

---

I'm extending the list of custom components with a lot of similary named and purposed `xx-section` classes:

```
/* components.css */

.pt-section {
  @apply pt-16 lg:pt-20;
}
.mt-section {
  @apply mt-16 lg:mt-20;
}
.-mt-section {
  @apply -mt-16 lg:-mt-20;
}


.top-section {
  @apply top-16 lg:top-20;
}
.-top-section {
  @apply -top-16 lg:-top-20;
}

/* 
Also: pb-section, mb-section, -mb-section bottom-section, -bottom-section, gap-section...
*/
```

The purpose of these utils is to be able to be able to quickly assign the same _responsive_ size to the different props of different elements. I hope its clear.

I see that it would be much nicer to define a single _responsive_ value in the TW config. I already do that for non-responsive constants describing `body` and `header` heights:

```
// tailwind.config.js

theme: {
  extend: {
    spacing: {
      "13": "3.25rem",
      "18": "4.5rem",
      "header": "5rem",
      "header-1": "calc(5rem - 1px)",
      "body": "calc(100vh - 5rem)",

      "sectionT": ??? dunno how to define, should be responsive !!!
      "sectionB": ??? dunno how to define, should be responsive !!!
    }
  }
},
```

In this way TW would auto-derive all that I need. I would get `mt-sectionT`, `gap-sectionB` etc. classes automatically.

Is there a way to declare custom responsive (breakpoint-depending) values in the `tailwind.config.js`? Without writing a custom plugin, I mean.

It sounds like a common case, yet I can't find the answer in the docs (checked https://tailwindcss.com/docs/theme#customizing-the-default-theme and other pages) or on StackOverflow.

**Comment by wongjn** [comment, marked as answer]
We can use `clamp()` to go between two values: `clamp(50px, (100vw - 1024px) * 9999, 100px)`. When `100vw - 1024px` is negative or `0`, it would be `50px` and when it is positive, it will be multiplied to a large number by `9999` and thus be clamped to `100px` instead. So for your configuration, it could look something like:
```js
module.exports = {
  theme: {
    extend: {
      spacing: ({ theme }) => ({
        "13": "3.25rem",
        "18": "4.5rem",
        "header": "5rem",
        "header-1": "calc(5rem - 1px)",
        "body": "calc(100vh - 5rem)",
        section: `clamp(1rem, (100vw - ${theme('screens.lg')}) * 9999, 1.25rem)`,
      }),
    },
  },
};
```

**Reply by ivan-kleshnin** [reply]
It's a smart hack, Justin! 👍 💡 Have you tried this approach in prod?

**Reply by wongjn** [reply]
No, I prefer sticking to the core classes where possible, so I'd use `pt-16 lg:pt-20` classes instead of abstracting it at the configuration level.
